doc/contributing/coding-conventions.xml

```diff
@@ -178,6 +178,12 @@ args.stdenv.mkDerivation (args // {
 </programlisting>
     </para>
    </listitem>
+   <listitem>
+    <para>
+     Arguments should be listed in the order they are used, with the
```

Comment:
Whats wrong with alphabetically?

Reply:
For one, the fact that a significant amount of the packages in nixpkgs already list their arguments in roughly the order that they're used, which would necessitate a lot of changes. Can you explain why alphabetical ordering is useful? I'm failing to see any particularly compelling reasons.

Reply:
Sorting alphabetically is easier to automate and in the end it does not matter which way we do it.

Reply:
Eh, I'm really not in favour of arbitrary decisions like that. It definitely does matter: code is for humans to read, not just a blob of data to be processed by tools. Ultimately, we're building NixOS to make our lives easier: code formatting is just another aspect of this.

I also think you overestimate how difficult it would be to implement: it's a simple matter of looping through the AST, grabbing bindings as they're used, and skipping duplicates. We'll be wanting to scan through the AST regardless, since it would be good to have a lint for unused arguments.

Reply:
Especially for more complex nix files it is way easier for me to just search through something sorted alphabetically instead something which seems to be sorted arbitrarily. phases and inputs are not sorted currently so the inputs could be random as well.

Reply:
Literally nobody would want drv attrs in alphabetical order. There are [efforts](https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-ordering.md) to specify something sensible, however.

Reply:
I'm exceedingly curious about your workflow by now, though. Do you not use grep/your editor's search functionality?

Reply:
> Whats wrong with alphabetically?
> Sorting alphabetically is easier to automate and in the end it does not matter which way we do it.

This elevates "is it easy to describe, measure, and enforce" far over _actual utility_ of a policy.
Lexicographical ordering is the absolute lowest common denominator — the sole utility I _can_ see is that it would group parameters with common prefixes together, but that's a convention we can specify explicitly.

> Especially for more complex nix files it is way easier for me to just search through something sorted alphabetically instead something which seems to be sorted arbitrarily.

If you're interested in viewing things in alphabetic order, it's an editor command away, just like searching by substring.
I'm not particularly interested in destroying a more useful semantic ordering in favour of … people who don't want to learn to hit `/` or `Ctrl-F`, as far as I can tell.

It's _their_ loss too, because we're actively destroying a mechanism for making code more skimmable and harder to grok quickly.

> phases and inputs are not sorted currently so the inputs could be random as well.

That sounds worth specifying a convention for, thanks!

Reply:
> Do you not use grep/your editor's search functionality?

Why should I use search if I can just scroll up and instantly find something? 
Searching for me is only useful in very big files which I can't overlook in two pages.

And no, alphabetically sorting is not one command away. If I change inputs and first need to sort them than they will stay this way. Having a pre commit hook is also not an option for me because they caused me more trouble than good.